In [156]:
%matplotlib inline
import matplotlib
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

print("Import libraries successfully")
# Load HTML using Pandas. If on Colab, you must upload the HTML to your session.
df_original = pd.read_html("croke.html")
print("Import data html successfully")

Import libraries successfully
Import data html successfully


In [157]:
type(df_original), len(df_original)

(list, 1)

In [158]:
df_original[0]

No salt                                                              \
       Site           pKa             n           δlow          δhigh   
0      Asp2  3.61 ± 0.05  1.30 ± 0.21   38.64 ± 0.09   41.99 ± 0.06   
1    Glu13c  4.21 ± 0.06  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08   
2     Glu20  4.07 ± 0.04  0.93 ± 0.06   33.10 ± 0.05   36.74 ± 0.04   
3     Glu28  4.20 ± 0.06  0.96 ± 0.09   33.16 ± 0.08   36.92 ± 0.08   
4    Glu35d    4.17±0.06  0.93 ± 0.08   33.11 ± 0.07   36.69 ± 0.07   
5    Glu46c    4.21±0.06  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08   
6     His50  6.78 ± 0.04  1.08 ± 0.08  123.11 ± 0.02  125.05 ± 0.06   
7     Glu57  4.20 ± 0.05  0.94 ± 0.08   33.10 ± 0.07   36.87 ± 0.07   
8     Glu61  4.04 ± 0.27  0.86 ± 0.35  121.05 ± 0.07  121.70 ± 0.08   
9     Glu83  4.35 ± 0.05  1.07 ± 0.10   33.14 ± 0.06   36.71 ± 0.07   
10    Asp98  3.68 ± 0.06  0.87 ± 0.11   38.23 ± 0.10   41.56 ± 0.07   
11   Glu104  4.39 ± 0.13  1.12 ± 0.26   33.13 ± 0.15   36.74 ± 0.17   
12  Glu105c  4.21 ± 0.06  0.90 ± 0.08   33.10 ± 0.08   36.79 ± 0.08   
13  Glu110d  4.50 ± 0.09  0.98 ± 0.19   34.34 ± 0.06   36.71 ± 0.08   
14   Glu114  4.70 ± 0.05  0.86 ± 0.08   33.09 ± 0.06   36.97 ± 0.09   
15   Asp115  4.15 ± 0.05  0.76 ± 0.06   38.18 ± 0.06   41.69 ± 0.05   
16   Asp119  4.02 ± 0.07  0.69 ± 0.06  122.85 ± 0.08  125.87 ± 0.06   
17   Asp121  4.62 ± 0.06  0.97 ± 0.08   39.11 ± 0.03   41.43 ± 0.06   
18   Glu123  4.87 ± 0.05  0.77 ± 0.06   33.16 ± 0.05   36.95 ± 0.08   
19   Glu126  4.89 ± 0.06  0.97 ± 0.18   32.53 ± 0.08   36.72 ± 0.14   
20   Glu130  4.53 ± 0.06  0.85 ± 0.09   33.09 ± 0.09   36.82 ± 0.09   
21   Glu131  4.58 ± 0.05  0.93 ± 0.09   33.26 ± 0.06   36.84 ± 0.07   
22   Asp135  4.16 ± 0.07  0.70 ± 0.06   38.18 ± 0.08   41.74 ± 0.08   
23   Glu137  4.82 ± 0.16  0.56 ± 0.11  123.36 ± 0.11  125.77 ± 0.17   
24   Glu139  4.69 ± 0.04  0.97 ± 0.10   33.14 ± 0.06   36.99 ± 0.08   
25    α-CO  3.65 ± 0.03  1.10 ± 0.07  8.537 ± 0.006  8.081 ± 0.003   

     150 mM NaCl                                                          
             pKa             n           δlow          δhigh Resonance  
0   3.63 ± 0.04  1.00 ± 0.08  122.81 ± 0.05  124.92 ± 0.03    Cβ/Nb  
1   4.32 ± 0.09  1.42 ± 0.35   32.85 ± 0.14   36.24 ± 0.16       Cγ  
2   4.24 ± 0.04  0.90 ± 0.08   32.77 ± 0.19   36.49 ± 0.18       Cγ  
3   4.27 ± 0.04  0.85 ± 0.06   32.77 ± 0.06   36.50 ± 0.06       Cγ  
4   4.08 ± 0.13  0.88 ± 0.20   30.42 ± 0.41   36.39 ± 0.29       Cγ  
5   4.31 ± 0.10  0.86 ± 0.15   32.99 ± 0.18   36.39 ± 0.13       Cγ  
6   6.52 ± 0.07  1.52 ± 0.37  123.45 ± 0.02  125.05 ± 0.05         N  
7   4.33 ± 0.09  0.76 ± 0.11   32.89 ± 0.13   36.49 ± 0.13       Cγ  
8   3.97 ± 0.06  1.12 ± 0.22  121.35 ± 0.04  122.07 ± 0.02         N  
9   4.33 ± 0.03  0.91 ± 0.05   32.70 ± 0.05   36.32 ± 0.05       Cγ  
10  4.00 ± 0.03  1.44 ± 0.16   38.60 ± 0.05   41.01 ± 0.04       Cβ  
11  4.51 ± 0.11  1.01 ± 0.17   33.02 ± 0.12   36.36 ± 0.14       Cγ  
12  4.49 ± 0.17  1.04 ± 0.35   33.85 ± 0.18   36.33 ± 0.20       Cγ  
13  4.31 ± 0.07  0.85 ± 0.09   32.65 ± 0.12   36.36 ± 0.10       Cγ  
14  4.49 ± 0.09  0.64 ± 0.08   32.59 ± 0.13   36.58 ± 0.14       Cγ  
15  4.18 ± 0.09  0.94 ± 0.16   38.10 ± 0.12   41.11 ± 0.11       Cβ  
16  3.99 ± 0.02  0.77 ± 0.03  123.27 ± 0.03  126.20 ± 0.02         N  
17  4.39 ± 0.03  0.78 ± 0.04  117.42 ± 0.03  119.36 ± 0.03    Cβ/Nb  
18  4.69 ± 0.10  0.78 ± 0.12   32.69 ± 0.13   36.46 ± 0.16       Cγ  
19  4.55 ± 0.10  0.91 ± 0.16   32.03 ± 0.16   36.12 ± 0.19       Cγ  
20  4.40 ± 0.09  1.09 ± 0.22   33.05 ± 0.14   36.38 ± 0.15       Cγ  
21  4.71 ± 0.15  0.61 ± 0.12   32.80 ± 0.22   37.09 ± 0.27       Cγ  
22  4.05 ± 0.06  0.77 ± 0.07   37.88 ± 0.09   41.16 ± 0.07       Cβ  
23  4.42 ± 0.03  0.92 ± 0.

In [159]:
df = df_original[0]["No salt"]
df

Site           pKa             n           δlow          δhigh
0      Asp2  3.61 ± 0.05  1.30 ± 0.21   38.64 ± 0.09   41.99 ± 0.06
1    Glu13c  4.21 ± 0.06  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08
2     Glu20  4.07 ± 0.04  0.93 ± 0.06   33.10 ± 0.05   36.74 ± 0.04
3     Glu28  4.20 ± 0.06  0.96 ± 0.09   33.16 ± 0.08   36.92 ± 0.08
4    Glu35d    4.17±0.06  0.93 ± 0.08   33.11 ± 0.07   36.69 ± 0.07
5    Glu46c    4.21±0.06  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08
6     His50  6.78 ± 0.04  1.08 ± 0.08  123.11 ± 0.02  125.05 ± 0.06
7     Glu57  4.20 ± 0.05  0.94 ± 0.08   33.10 ± 0.07   36.87 ± 0.07
8     Glu61  4.04 ± 0.27  0.86 ± 0.35  121.05 ± 0.07  121.70 ± 0.08
9     Glu83  4.35 ± 0.05  1.07 ± 0.10   33.14 ± 0.06   36.71 ± 0.07
10    Asp98  3.68 ± 0.06  0.87 ± 0.11   38.23 ± 0.10   41.56 ± 0.07
11   Glu104  4.39 ± 0.13  1.12 ± 0.26   33.13 ± 0.15   36.74 ± 0.17
12  Glu105c  4.21 ± 0.06  0.90 ± 0.08   33.10 ± 0.08   36.79 ± 0.08
13  Glu110d  4.50 ± 0.09  0.98 ± 0.19   34.34 ± 0.06   36.71 ± 0.08
14   Glu114  4.70 ± 0.05  0.86 ± 0.08   33.09 ± 0.06   36.97 ± 0.09
15   Asp115  4.15 ± 0.05  0.76 ± 0.06   38.18 ± 0.06   41.69 ± 0.05
16   Asp119  4.02 ± 0.07  0.69 ± 0.06  122.85 ± 0.08  125.87 ± 0.06
17   Asp121  4.62 ± 0.06  0.97 ± 0.08   39.11 ± 0.03   41.43 ± 0.06
18   Glu123  4.87 ± 0.05  0.77 ± 0.06   33.16 ± 0.05   36.95 ± 0.08
19   Glu126  4.89 ± 0.06  0.97 ± 0.18   32.53 ± 0.08   36.72 ± 0.14
20   Glu130  4.53 ± 0.06  0.85 ± 0.09   33.09 ± 0.09   36.82 ± 0.09
21   Glu131  4.58 ± 0.05  0.93 ± 0.09   33.26 ± 0.06   36.84 ± 0.07
22   Asp135  4.16 ± 0.07  0.70 ± 0.06   38.18 ± 0.08   41.74 ± 0.08
23   Glu137  4.82 ± 0.16  0.56 ± 0.11  123.36 ± 0.11  125.77 ± 0.17
24   Glu139  4.69 ± 0.04  0.97 ± 0.10   33.14 ± 0.06   36.99 ± 0.08
25    α-CO  3.65 ± 0.03  1.10 ± 0.07  8.537 ± 0.006  8.081 ± 0.003

## Task 1
The above table cannot be readily processed for various reasons. 

- Firstly, the numbers are actually strings and contain the unicode character ±.
- Secondly, values and errors are mixed. 


In [160]:
type(df['Site'])

pandas.core.series.Series

In [161]:
df.pKa.str.split("±").str.get(0)

0     3.61 Â
1     4.21 Â
2     4.07 Â
3     4.20 Â
4      4.17Â
5      4.21Â
6     6.78 Â
7     4.20 Â
8     4.04 Â
9     4.35 Â
10    3.68 Â
11    4.39 Â
12    4.21 Â
13    4.50 Â
14    4.70 Â
15    4.15 Â
16    4.02 Â
17    4.62 Â
18    4.87 Â
19    4.89 Â
20    4.53 Â
21    4.58 Â
22    4.16 Â
23    4.82 Â
24    4.69 Â
25    3.65 Â
Name: pKa, dtype: object

In [162]:
df[['Site']]

Site
0      Asp2
1    Glu13c
2     Glu20
3     Glu28
4    Glu35d
5    Glu46c
6     His50
7     Glu57
8     Glu61
9     Glu83
10    Asp98
11   Glu104
12  Glu105c
13  Glu110d
14   Glu114
15   Asp115
16   Asp119
17   Asp121
18   Glu123
19   Glu126
20   Glu130
21   Glu131
22   Asp135
23   Glu137
24   Glu139
25    α-CO

In [163]:
df = df.rename(columns={'δlow': 'low',
                   'δhigh': 'high'})
df

Site           pKa             n             low            high
0      Asp2  3.61 ± 0.05  1.30 ± 0.21   38.64 ± 0.09   41.99 ± 0.06
1    Glu13c  4.21 ± 0.06  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08
2     Glu20  4.07 ± 0.04  0.93 ± 0.06   33.10 ± 0.05   36.74 ± 0.04
3     Glu28  4.20 ± 0.06  0.96 ± 0.09   33.16 ± 0.08   36.92 ± 0.08
4    Glu35d    4.17±0.06  0.93 ± 0.08   33.11 ± 0.07   36.69 ± 0.07
5    Glu46c    4.21±0.06  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08
6     His50  6.78 ± 0.04  1.08 ± 0.08  123.11 ± 0.02  125.05 ± 0.06
7     Glu57  4.20 ± 0.05  0.94 ± 0.08   33.10 ± 0.07   36.87 ± 0.07
8     Glu61  4.04 ± 0.27  0.86 ± 0.35  121.05 ± 0.07  121.70 ± 0.08
9     Glu83  4.35 ± 0.05  1.07 ± 0.10   33.14 ± 0.06   36.71 ± 0.07
10    Asp98  3.68 ± 0.06  0.87 ± 0.11   38.23 ± 0.10   41.56 ± 0.07
11   Glu104  4.39 ± 0.13  1.12 ± 0.26   33.13 ± 0.15   36.74 ± 0.17
12  Glu105c  4.21 ± 0.06  0.90 ± 0.08   33.10 ± 0.08   36.79 ± 0.08
13  Glu110d  4.50 ± 0.09  0.98 ± 0.19   34.34 ± 0.06   36.71 ± 0.08
14   Glu114  4.70 ± 0.05  0.86 ± 0.08   33.09 ± 0.06   36.97 ± 0.09
15   Asp115  4.15 ± 0.05  0.76 ± 0.06   38.18 ± 0.06   41.69 ± 0.05
16   Asp119  4.02 ± 0.07  0.69 ± 0.06  122.85 ± 0.08  125.87 ± 0.06
17   Asp121  4.62 ± 0.06  0.97 ± 0.08   39.11 ± 0.03   41.43 ± 0.06
18   Glu123  4.87 ± 0.05  0.77 ± 0.06   33.16 ± 0.05   36.95 ± 0.08
19   Glu126  4.89 ± 0.06  0.97 ± 0.18   32.53 ± 0.08   36.72 ± 0.14
20   Glu130  4.53 ± 0.06  0.85 ± 0.09   33.09 ± 0.09   36.82 ± 0.09
21   Glu131  4.58 ± 0.05  0.93 ± 0.09   33.26 ± 0.06   36.84 ± 0.07
22   Asp135  4.16 ± 0.07  0.70 ± 0.06   38.18 ± 0.08   41.74 ± 0.08
23   Glu137  4.82 ± 0.16  0.56 ± 0.11  123.36 ± 0.11  125.77 ± 0.17
24   Glu139  4.69 ± 0.04  0.97 ± 0.10   33.14 ± 0.06   36.99 ± 0.08
25    α-CO  3.65 ± 0.03  1.10 ± 0.07  8.537 ± 0.006  8.081 ± 0.003

In [164]:
type(df['pKa'][0])

str

In [165]:
df[['pKa', 'err']] = df['pKa'].str.split(r'\s*±\s*',expand=True).astype(float)

In [166]:
df['pKa'][0]

np.float64(3.61)

In [167]:
df

Site   pKa             n             low            high   err
0      Asp2  3.61  1.30 ± 0.21   38.64 ± 0.09   41.99 ± 0.06  0.05
1    Glu13c  4.21  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08  0.06
2     Glu20  4.07  0.93 ± 0.06   33.10 ± 0.05   36.74 ± 0.04  0.04
3     Glu28  4.20  0.96 ± 0.09   33.16 ± 0.08   36.92 ± 0.08  0.06
4    Glu35d  4.17  0.93 ± 0.08   33.11 ± 0.07   36.69 ± 0.07  0.06
5    Glu46c  4.21  0.90 ± 0.08   33.11 ± 0.08   36.79 ± 0.08  0.06
6     His50  6.78  1.08 ± 0.08  123.11 ± 0.02  125.05 ± 0.06  0.04
7     Glu57  4.20  0.94 ± 0.08   33.10 ± 0.07   36.87 ± 0.07  0.05
8     Glu61  4.04  0.86 ± 0.35  121.05 ± 0.07  121.70 ± 0.08  0.27
9     Glu83  4.35  1.07 ± 0.10   33.14 ± 0.06   36.71 ± 0.07  0.05
10    Asp98  3.68  0.87 ± 0.11   38.23 ± 0.10   41.56 ± 0.07  0.06
11   Glu104  4.39  1.12 ± 0.26   33.13 ± 0.15   36.74 ± 0.17  0.13
12  Glu105c  4.21  0.90 ± 0.08   33.10 ± 0.08   36.79 ± 0.08  0.06
13  Glu110d  4.50  0.98 ± 0.19   34.34 ± 0.06   36.71 ± 0.08  0.09
14   Glu114  4.70  0.86 ± 0.08   33.09 ± 0.06   36.97 ± 0.09  0.05
15   Asp115  4.15  0.76 ± 0.06   38.18 ± 0.06   41.69 ± 0.05  0.05
16   Asp119  4.02  0.69 ± 0.06  122.85 ± 0.08  125.87 ± 0.06  0.07
17   Asp121  4.62  0.97 ± 0.08   39.11 ± 0.03   41.43 ± 0.06  0.06
18   Glu123  4.87  0.77 ± 0.06   33.16 ± 0.05   36.95 ± 0.08  0.05
19   Glu126  4.89  0.97 ± 0.18   32.53 ± 0.08   36.72 ± 0.14  0.06
20   Glu130  4.53  0.85 ± 0.09   33.09 ± 0.09   36.82 ± 0.09  0.06
21   Glu131  4.58  0.93 ± 0.09   33.26 ± 0.06   36.84 ± 0.07  0.05
22   Asp135  4.16  0.70 ± 0.06   38.18 ± 0.08   41.74 ± 0.08  0.07
23   Glu137  4.82  0.56 ± 0.11  123.36 ± 0.11  125.77 ± 0.17  0.16
24   Glu139  4.69  0.97 ± 0.10   33.14 ± 0.06   36.99 ± 0.08  0.04
25    α-CO  3.65  1.10 ± 0.07  8.537 ± 0.006  8.081 ± 0.003  0.03

In [168]:
df[['resname', 'resnum']] = df['Site'].str.extract(r'([A-Za-z]+)(\d+)')

In [169]:
df = df[['pKa', 'err', 'resname', 'resnum']]
df

pKa   err resname resnum
0   3.61  0.05     Asp      2
1   4.21  0.06     Glu     13
2   4.07  0.04     Glu     20
3   4.20  0.06     Glu     28
4   4.17  0.06     Glu     35
5   4.21  0.06     Glu     46
6   6.78  0.04     His     50
7   4.20  0.05     Glu     57
8   4.04  0.27     Glu     61
9   4.35  0.05     Glu     83
10  3.68  0.06     Asp     98
11  4.39  0.13     Glu    104
12  4.21  0.06     Glu    105
13  4.50  0.09     Glu    110
14  4.70  0.05     Glu    114
15  4.15  0.05     Asp    115
16  4.02  0.07     Asp    119
17  4.62  0.06     Asp    121
18  4.87  0.05     Glu    123
19  4.89  0.06     Glu    126
20  4.53  0.06     Glu    130
21  4.58  0.05     Glu    131
22  4.16  0.07     Asp    135
23  4.82  0.16     Glu    137
24  4.69  0.04     Glu    139
25  3.65  0.03     NaN    NaN

In [172]:
type(df['resnum'][0])

str

In [176]:
df.loc[:,'resnum'] = pd.to_numeric(df['resnum'], errors='coerce').astype("Int64")
df

pKa   err resname resnum
0   3.61  0.05     Asp      2
1   4.21  0.06     Glu     13
2   4.07  0.04     Glu     20
3   4.20  0.06     Glu     28
4   4.17  0.06     Glu     35
5   4.21  0.06     Glu     46
6   6.78  0.04     His     50
7   4.20  0.05     Glu     57
8   4.04  0.27     Glu     61
9   4.35  0.05     Glu     83
10  3.68  0.06     Asp     98
11  4.39  0.13     Glu    104
12  4.21  0.06     Glu    105
13  4.50  0.09     Glu    110
14  4.70  0.05     Glu    114
15  4.15  0.05     Asp    115
16  4.02  0.07     Asp    119
17  4.62  0.06     Asp    121
18  4.87  0.05     Glu    123
19  4.89  0.06     Glu    126
20  4.53  0.06     Glu    130
21  4.58  0.05     Glu    131
22  4.16  0.07     Asp    135
23  4.82  0.16     Glu    137
24  4.69  0.04     Glu    139
25  3.65  0.03     NaN   <NA>

In [177]:
df.to_csv("croke.csv", index=False)

## Task 3

The acid dissocitation constant of an isolated glutamic acid ("Glu") is p =4.1. As a function of residue number, plot the shift from this reference value for all glutamates in the above tabel. Hints: Use matplotlib and a pandas mask using e.g. mask = df.resname == "Glu".